File: seminars/sem8.fs
type 'a Tree =
  EmptyTree
  | Node of 'a * 'a Tree * 'a Tree

let singleton x = Node (x, EmptyTree, EmptyTree)

let rec treeInsert x = function
  EmptyTree -> singleton x 
  | Node (a, left, right) -> 
    if x = a then Node (x, left, right) 
    else 
      if x < a then Node (a, (treeInsert x left), right) 
      else Node (a, left, (treeInsert x right))
// when 'a : comparsion

let list2tree list =
 let rec l2t acc = function
   [] -> acc
   | (head::tail) -> l2t (treeInsert head acc) tail
 in l2t EmptyTree list

// list2tree через fold

list2tree [12; 1; 6; 4; 90; 9]

tree2list // сами


let treesort x = x |> list2tree |> tree2list

treesort [12; 1; 6; 4; 90; 9]

list2tree [12; 12; 12; 13; 13; 14]
Как будет выглядеть дерево?

type 'a Tree =
  EmptyTree
  | Node of 'a * int * 'a Tree * 'a Tree

let singleton x = ? 

let rec treeInsert x = ?

let list2tree x = 
  let rec l2t acc = function
    | [] -> acc
    | (head::tail) -> l2t (treeInsert head acc) tail
l2t EmptyTree x

let foldTree treeFunction listValue tree =
    let rec loop tree cont =
        match tree with
        | EmptyTree -> cont listValue
        | Node (x, left, right) -> loop left (fun leftAcc -> 
            loop right (fun rightAcc -> 
              cont (treeFunction x leftAcc rightAcc)
            )
          )
    loop tree (fun x -> x)
    
// написать foldTree без продолжений


let sumTree = foldTree (fun x left right -> x + left + right) 0
[2;7;4;3;5;8] |> list2tree |> sumTree

let heightTree = foldTree (fun _ left right -> 1 + max left right) 0
[2;7;4;3;5;8] |> list2tree |> heightTree 

let tree2List = foldTree (fun x left right -> left @ (x :: right)) []
[2;7;4;3;5;8] |> list2tree |> tree2List

// найти максимальное значение в дереве

// перевернуть дерево

open System

let generate = 
  let rnd = Random()
  rnd.Next(42)

generate 0 42

// сгенерировать дерево

// проверить, что два дерева подобны

type 'a Tree =
  | EmptyTree
  | Leaf of 'a
  | Node of 'a * 'a Tree list

Node(1, [Leaf 3; EmptyTree; Node(2, [])])

// найти высоту дерева, EmptyTree не считаются

type Tree<'LeafData,'INodeData> =
    | LeafNode of 'LeafData
    | InternalNode of 'INodeData * Tree<'LeafData,'INodeData> seq

let rec fold fLeaf fNode acc (tree:Tree<'LeafData,'INodeData>) : 'r = 
    let recurse = fold fLeaf fNode  
    match tree with
    | LeafNode leafInfo -> 
        fLeaf acc leafInfo 
    | InternalNode (nodeInfo, subtrees) -> 
        Seq.fold recurse (fNode acc nodeInfo) subtrees 

// map через fold?

let rec map fLeaf fNode (tree:Tree<'LeafData,'INodeData>) = ?

type FileInfo = {name:string; fileSize:int}
type DirectoryInfo = {name:string; dirSize:int}

type FileSystemItem = Tree<FileInfo,DirectoryInfo>

let fromFile (fileInfo:FileInfo) = 
    LeafNode fileInfo 

let fromDir (dirInfo:DirectoryInfo) subitems = 
    InternalNode (dirInfo,subitems)

let readme = fromFile {name="readme.txt"; fileSize=1}
let config = fromFile {name="config.json"; fileSize=2}
let build  = fromFile {name="build.sh"; fileSize=3}
let src = fromDir {name="src"; dirSize=10} [readme; config; build]
let bin = fromDir {name="bin"; dirSize=10} []
let root = fromDir {name="root"; dirSize=5} [src; bin]

let totalSize fileSystemItem =
    let fFile acc (file:FileInfo) = 
        acc + file.fileSize
    let fDir acc (dir:DirectoryInfo)= 
        acc + dir.dirSize
    fold fFile fDir 0 fileSystemItem 

readme |> totalSize  
src |> totalSize     
root |> totalSize    

// largestFile : fileSystemItem:Tree<FileInfo,'a> -> FileInfo option
let largestFile fileSystemItem =

readme |> largestFile
src |> largestFile
bin |> largestFile
root |> largestFile

open System
open System.IO

DirectoryInfo("/home/und/fsharp")

type FileSystemTree = Tree<FileInfo,DirectoryInfo>

let fromFile (fileInfo:FileInfo) = 
    LeafNode fileInfo 

let rec fromDir (dirInfo:DirectoryInfo) = 
    let subItems = seq {
        yield! dirInfo.EnumerateFiles() |> Seq.map fromFile
        yield! dirInfo.EnumerateDirectories() |> Seq.map fromDir
    }
    InternalNode (dirInfo,subItems)

let totalSize fileSystemItem =
    let fFile acc (file:FileInfo) = 
        acc + file.Length
    let fDir acc (dir:DirectoryInfo)= 
        acc 
    fold fFile fDir 0L fileSystemItem 
   
let currentDir = fromDir (DirectoryInfo("/home/und/fsharp"))

currentDir |> totalSize  

let largestFile fileSystemItem =
    let fFile (largestSoFarOpt:FileInfo option) (file:FileInfo) = 
        match largestSoFarOpt with
        | None -> 
            Some file                
        | Some largestSoFar -> 
            if largestSoFar.Length > file.Length then
                Some largestSoFar
            else
                Some file

    let fDir largestSoFarOpt dirInfo = 
        largestSoFarOpt

    fold fFile fDir None fileSystemItem

currentDir |> largestFile  

let dirListing fileSystemItem =
    let printDate (d:DateTime) = d.ToString()
    let mapFile (fi:FileInfo) = 
        sprintf "%10i  %s  %-s"  fi.Length (printDate fi.LastWriteTime) fi.Name
    let mapDir (di:DirectoryInfo) = 
        di.FullName 
    map mapFile mapDir fileSystemItem

currentDir 
    |> dirListing 
    |> map (printfn "%s") (printfn "\n%s")


// filter fs, fsx files

File: seminars/Sem8.hs
module Sem8 where

data BinTree a
    = EmptyTree
    | Node a  (BinTree a) (BinTree a)
    deriving (Show, Ord)


singleton :: a -> BinTree a
singleton x = Node x EmptyTree EmptyTree


treeInsert :: Ord a => a -> BinTree a -> BinTree a
treeInsert x EmptyTree = singleton x
treeInsert x (Node k left right)
    | x == k    = Node x left right
    | x < k     = Node k (treeInsert x left) right
    | otherwise = Node k left (treeInsert x right)


list2tree :: Ord a => [a] -> BinTree a
list2tree = foldl (flip treeInsert) EmptyTree


tree2list :: BinTree a -> [a]
tree2list EmptyTree         = []
tree2list (Node x left right) = tree2list left ++ [x] ++ tree2list right

treeSort :: Ord a => [a] -> [a]
treeSort = tree2list . list2tree

-- not a good way to make a long lambda

-- foldTree :: (t -> t1 -> t1 -> t1) -> t1 -> BinTree t -> t1
-- foldTree treeFunction listValue tree = loop tree id where
--     loop EmptyTree cont           = cont listValue
--     loop (Node x left right) cont = loop left (\leftAcc ->
--         loop right (\rightAcc ->
--             cont (treeFunction x leftAcc rightAcc)
--             )
--         )


foldBinTree :: (t -> t1 -> t1 -> t1) -> t1 -> BinTree t -> t1
foldBinTree treeFunction listValue = fold' where
    fold' EmptyTree           = listValue
    fold' (Node x left right) = treeFunction x (fold' left) (fold' right)


sumTree :: Num a => BinTree a -> a
sumTree = foldBinTree (\x a b -> x + a + b) 0

instance (Eq a) => Eq (BinTree a) where
    EmptyTree == EmptyTree                           = True
    EmptyTree == Node{}                              = False
    Node{} == EmptyTree                              = False
    (Node x1 left1 right1) == (Node x2 left2 right2) = x1 == x2 &&
                                                        left1 == left2 &&
                                                         right1 == right2


data Tree tLeaf tNode =
      LeafNode tLeaf
    | InternalNode tNode [Tree tLeaf tNode]


foldTree :: (t -> tleaf -> t) -> (t -> tnode -> t) -> t -> Tree tleaf tnode -> t
foldTree fLeaf fNode = fold' where
    fold' acc tree = case tree of
        LeafNode leafInfo              ->
            fLeaf acc leafInfo
        InternalNode nodeInfo subTrees ->
            foldl fold' (fNode acc nodeInfo) subTrees


mapTree :: (tLeaf -> tLeaf') -> (tNode -> tNode') -> Tree tLeaf tNode -> Tree tLeaf' tNode'
mapTree fLeaf fNode = mapTree' where
    mapTree' (LeafNode leafInfo)              = LeafNode $ fLeaf leafInfo
    mapTree' (InternalNode nodeInfo subTrees) =
        InternalNode (fNode nodeInfo) (map mapTree' subTrees)
